ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,308 @@
+import re
+from typing import List, Union
+
+import numpy as np
+import tables
+from astropy.table import join, vstack, Table
+
+from ..core import Component, traits
+from ..instrument import SubarrayDescription, TelescopeDescription
+from .astropy_helpers import read_table
+
+__all__ = ["get_tel_ids", "get_structure", "TableLoader"]
+
+PARAMETERS_GROUP = "/dl1/event/telescope/parameters"
+IMAGES_GROUP = "/dl1/event/telescope/images"
+GEOMETRY_GROUP = "/dl2/event/subarray/geometry"
+TRIGGER_TABLE = "/dl1/event/subarray/trigger"
+SHOWER_TABLE = "/simulation/event/subarray/shower"
+TRUE_IMAGES_GROUP = "/simulation/event/telescope/images"
+
+by_id_RE = re.compile(r"tel_\d+")
+
+
+def get_tel_ids(
+    subarray: SubarrayDescription,
+    telescopes: List[Union[int, str, TelescopeDescription]]
+) -> List[int]:
+    ids = set()
+
+    for telescope in telescopes:
+        if isinstance(telescope, int):
+            ids.add(telescope)
+        ids.update(subarray.get_tel_ids_for_type(telescope))
+
+    return sorted(ids)
+
+
+def get_structure(h5file):
+
+    if PARAMETERS_GROUP in h5file:
+        g = h5file.root[PARAMETERS_GROUP]
+    elif IMAGES_GROUP in h5file:
+        g = h5file.root[IMAGES_GROUP]
+    else:
+        raise ValueError(f"No DL1 parameters data in h5file: {h5file}")
+
+    key = next(iter(g._v_children))
+    if by_id_RE.fullmatch(key):
+        return "by_id"
+    else:
+        return "by_type"
+
+
+class TableLoader(Component):
+    """ A helper class to load and join tables from a DL1 file"""
+
+    input_url = traits.Path(directory_ok=False, exists=True).tag(config=True)
+
+    load_dl1_images = traits.Bool(False).tag(config=True)
+    load_dl1_parameters = traits.Bool(True).tag(config=True)
+    load_dl2_geometry = traits.Bool(False).tag(config=True)
+    load_simulated = traits.Bool(False).tag(config=True)
+    load_true_images = traits.Bool(False).tag(config=True)
+    load_trigger = traits.Bool(True).tag(config=True)
+    load_instrument = traits.Bool(False).tag(config=True)
+
+    def __init__(self, input_url=None, **kwargs):
+        # enable using input_url as posarg
+        if input_url not in {None, traits.Undefined}:
+            kwargs["input_url"] = input_url
+        super().__init__(**kwargs)
+
+        self.subarray = SubarrayDescription.from_hdf(self.input_url)
+        self.h5file = tables.open_file(self.input_url, mode="r")
```

Comment:
should add the following after this line:

```py3
Provenance().add_input_file(self.input_url, role="Event data") 
```

Reply:
added